x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/sort/BytesRefBucketedSort.java

```diff
@@ -0,0 +1,392 @@
+        this.breaker = breaker;
+        this.label = label;
+        common = new BucketedSortCommon(bigArrays, order, bucketSize);
+        boolean success = false;
+        try {
+            values = bigArrays.newObjectArray(0);
+            success = true;
+        } finally {
+            if (success == false) {
+                close();
+            }
+        }
+    }
+
+    private void checkInvariant(int bucket) {
+        if (Assertions.ENABLED == false) {
+            return;
+        }
+        long rootIndex = common.rootIndex(bucket);
+        long requiredSize = common.requiredSize(rootIndex);
+        if (values.size() < requiredSize) {
+            throw new AssertionError("values too short " + values.size() + " < " + requiredSize);
+        }
+        if (values.get(rootIndex) == null) {
+            throw new AssertionError("new gather offset can't be null");
+        }
+        if (common.inHeapMode(bucket) == false) {
+            common.assertValidNextOffset(getNextGatherOffset(rootIndex));
+        } else {
+            for (long l = rootIndex; l < common.endIndex(rootIndex); l++) {
+                if (values.get(rootIndex) == null) {
+                    throw new AssertionError("values missing in heap mode");
+                }
+            }
+        }
+    }
+
+    /**
+     * Collects a {@code value} into a {@code bucket}.
+     * <p>
+     *     It may or may not be inserted in the heap, depending on if it is better than the current root.
+     * </p>
+     */
+    public void collect(BytesRef value, int bucket) {
+        long rootIndex = common.rootIndex(bucket);
+        if (common.inHeapMode(bucket)) {
+            if (betterThan(value, values.get(rootIndex).bytesRefView())) {
+                clearedBytesAt(rootIndex).append(value);
+                downHeap(rootIndex, 0);
+            }
+            checkInvariant(bucket);
+            return;
+        }
+        // Gathering mode
+        long requiredSize = common.requiredSize(rootIndex);
+        if (values.size() < requiredSize) {
+            grow(requiredSize);
+        }
+        int next = getNextGatherOffset(rootIndex);
+        common.assertValidNextOffset(next);
+        long index = next + rootIndex;
+        clearedBytesAt(index).append(value);
+        if (next == 0) {
+            common.enableHeapMode(bucket);
+            heapify(rootIndex);
+        } else {
+            ByteUtils.writeIntLE(next - 1, values.get(rootIndex).bytes(), 0);
+        }
+        checkInvariant(bucket);
+    }
+
+    /**
+     * Merge the values from {@code other}'s {@code otherGroupId} into {@code groupId}.
+     */
+    public void merge(int bucket, BytesRefBucketedSort other, int otherBucket) {
+        long otherRootIndex = other.common.rootIndex(otherBucket);
+        if (otherRootIndex >= other.values.size()) {
+            // The value was never collected.
+            return;
+        }
+        other.checkInvariant(bucket);
```

Comment:
We're already checking this on collect(), what could go wrong here?

Reply:
In my tests I had screwed up *something* and the reads were blowing up in interesting ways. Adding this made it a lot more obvious what I'd broken. I forget what it was. But I believe this is NOOPed at runtime outside of tests so we're probably safe keeping it. If just for paranoia.